go-controller/pkg/ovn/ovn.go

```diff
@@ -225,30 +242,89 @@ func getPodNamespacedName(pod *kapi.Pod) string {
+	// database in large clusters. ovn-controllers should be upgraded to a version
+	// that supports them before the option is turned on by the master.
+	nbGlobal := nbdb.NBGlobal{
+		Options: map[string]string{"use_logical_dp_groups": "true"},
+	}
+	if err := libovsdbops.UpdateNBGlobalSetOptions(mc.nbClient, &nbGlobal); err != nil {
+		return fmt.Errorf("failed to set NB global option to enable logical datapath groups: %v", err)
+	}
+
+	// Start and sync the watch factory to begin listening for events
+	if err := mc.watchFactory.Start(); err != nil {
+		return err
+	}
+
+	metrics.RunTimestamp(mc.stopChan, mc.sbClient, mc.nbClient)
+	metrics.MonitorIPSec(mc.nbClient)
+	if config.Metrics.EnableConfigDuration {
+		// with k=10,
+		//  for a cluster with 10 nodes, measurement of 1 in every 100 requests
+		//  for a cluster with 100 nodes, measurement of 1 in every 1000 requests
+		metrics.GetConfigDurationRecorder().Run(mc.nbClient, mc.kube, 10, time.Second*5, mc.stopChan)
+	}
+
+	mc.podRecorder.Run(mc.sbClient, mc.stopChan)
+	return nil
+}
+
 // NewOvnController creates a new OVN controller for creating logical network
 // infrastructure and policy
-func NewOvnController(ovnClient *util.OVNClientset, wf *factory.WatchFactory, stopChan <-chan struct{}, addressSetFactory addressset.AddressSetFactory,
-	libovsdbOvnNBClient libovsdbclient.Client, libovsdbOvnSBClient libovsdbclient.Client,
-	recorder record.EventRecorder) *Controller {
+func (mc *OvnMHController) NewOvnController(addressSetFactory addressset.AddressSetFactory) (*Controller, error) {
+	clusterIPNet, err := config.ParseClusterSubnetEntries(config.Default.RawClusterSubnets)
+	if err != nil {
+		return nil, fmt.Errorf("cluster subnet %s is invalid: %v", config.Default.RawClusterSubnets, err)
+	}
+
 	if addressSetFactory == nil {
-		addressSetFactory = addressset.NewOvnAddressSetFactory(libovsdbOvnNBClient)
+		addressSetFactory = addressset.NewOvnAddressSetFactory(mc.nbClient)
 	}
-	svcController, svcFactory := newServiceController(ovnClient.KubeClient, libovsdbOvnNBClient, recorder)
-	egressSvcController := newEgressServiceController(ovnClient.KubeClient, libovsdbOvnNBClient, svcFactory, stopChan)
+	svcController, svcFactory := newServiceController(mc.client, mc.nbClient, mc.recorder)
+	egressSvcController := newEgressServiceController(mc.client, mc.nbClient, svcFactory, mc.stopChan)
 	var hybridOverlaySubnetAllocator *subnetallocator.SubnetAllocator
 	if config.HybridOverlay.Enabled {
 		hybridOverlaySubnetAllocator = subnetallocator.NewSubnetAllocator()
 	}
-	return &Controller{
-		client: ovnClient.KubeClient,
-		kube: &kube.Kube{
-			KClient:              ovnClient.KubeClient,
-			EIPClient:            ovnClient.EgressIPClient,
-			EgressFirewallClient: ovnClient.EgressFirewallClient,
-			CloudNetworkClient:   ovnClient.CloudNetworkClient,
-		},
-		watchFactory:                 wf,
+
+	stopChan := mc.stopChan
+	oc := &Controller{
```

Comment:
will each controller really have its own retry mechanism per resource? How will egress IPs and other features work with secondary networks? Perhaps some of these can be restricted to only the default controller?

Reply:
This is a really good point.

I would prefer to be on the safe side, and restrict some features to only the default controller. Some examples: 
- services (**eventually** we'd also cover this for secondary networks).
- egress IP
- ...